### ISSUE TITLE
Sparse matrix inverses in SR give wrong results on Sage 9.8

### ISSUE THREAD

**Issue by pvir** [sagemath/sage#35653]
### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### Did you read the documentation and troubleshoot guide?

- [X] I have read the documentation and troubleshoot guide

### Environment

```markdown
- **OS**: Redhat / Mambaforge / conda-forge (`mamba install sage=9.8`)
- **Sage Version**: 9.8
```

### Steps To Reproduce

```
sage: var('x')
x
sage: diagonal_matrix([x]).inverse()
[1]
```
which is not quite right. Inverses seem to be also more generally broken for sparse matrices:
```
sage: M = MatrixSpace(SR,2,2,sparse=True)
sage: M([[x,0],[0,x]]).inverse()
[1 0]
[0 1]
sage: M([[x,1],[0,x]]).inverse()
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
```

Running `sage -t .` in src/sage/matrix also generates several errors, [sage-matrix-test.log](https://github.com/sagemath/sage/files/11514318/sage-matrix-test.log) 

I can also reproduce this on https://cocalc.com so it's probably not specific to this installation.

### Expected Behavior

Correct results.

### Actual Behavior

Matrix inverses give wrong results.

### Additional Information

Sage 9.7 gives correct results, so this looks like a regression.

**Comment by DaveWitteMorris** [comment]
Thanks for the bug report.  I confirm that `diagonal_matrix([x]).inverse()` is wrong for symbolic matrices.  On the other hand, I am not getting any doctest failures from `sage -t` (and the doctests are checked before every release), so there may be something wrong with your installation.

Diagnosis of the bug: The `__invert__` method in src/sage/matrix/matrix0.pyx calls the `echelonize` method of src/sage/matrix/matrix2.pyx.  This method decides to use the "scaled_partial_pivoting" algorithm (because `SR` is not an exact ring, and `((0.42)).abs()` does not raise an error).  But that algorithm should only be applied to numeric matrices, not matrices over `SR`, so we get nonsense.

There is no good general method to invert matrices over `SR`, so I'm not sure what we should do, but perhaps the `echelonize` method should revert to choosing the "classical" algorithm for symbolic matrices.

**Comment by DaveWitteMorris** [comment]
@dimpase: Do you have a suggestion for how to fix this?  I think you wrote the `((0.42)).abs()` test, but I think it needs to be changed, because I don't think matrices over `SR` should be using the "scaled_partial_pivoting" algorithm.

**Comment by dimpase** [comment]
is it for sparse only?

**Comment by DaveWitteMorris** [comment]
Yes, we get the correct inverse by specifying `sparse=False` in the definition of the matrix (and some other examples that I tried).

**Comment by dimpase** [comment]
the `0.42` test is in `echelonize`, so somehow it does not get there for non-sparse matrices.
It's in 24299ced2b38 - which had a special case test for `SR`, but this test is not in the current develop.

**Comment by dimpase** [comment]
matrix inversion is done via `echelonize()`, and here lies a strange discrepancy between sparse and dense:
```
sage: M = MatrixSpace(SR,2,2,sparse=True)
sage: m=M([[x,0],[0,x]])
sage: A = m.augment(m.parent().identity_matrix()); A
[x 0 1 0]
[0 x 0 1]
sage: A.echelonize() # wrong!
sage: A
[x 0 1 0]
[0 x 0 1]
sage: M = MatrixSpace(SR,2,2,sparse=False)
sage: m=M([[x,0],[0,x]])
sage: A = m.augment(m.parent().identity_matrix()); A
[x 0 1 0]
[0 x 0 1]
sage: A.echelonize() # right.
sage: A
[  1   0 1/x   0]
[  0   1   0 1/x]
```

**Comment by dimpase** [comment]
A way to fix this would be to create `src/sage/matrix/matrix_symbolic_sparse.pyx`,  basically copy-paste from `src/sage/matrix/matrix_symbolic_dense.pyx`.

That's quite ugly, of course. Perhaps better would be to have `src/sage/matrix/matrix_symbolic.pyx`, to handle both cases, but I'm not sure whether this will fit into the matrix classes structure.

**Comment by dimpase** [comment]
I have a branch to fix it,not yet working

**Comment by dimpase** [comment]
fix is ready on #35714 - please review

**Comment by DaveWitteMorris** [comment]
Fixed by [#35714](https://github.com/sagemath/sage/pull/35714).
